core/src/stream.rs

```diff
@@ -13,14 +13,59 @@ const TIMEOUT: Duration = Duration::from_millis(10_000);
 #[cfg(not(test))]
 const TIMEOUT: Duration = Duration::from_millis(1000);
 
-/// Represents messsage used by the stream
+/// Error type with generic for actual Stream/Sink error type
+#[derive(thiserror::Error, Debug)]
+pub enum Error<InternalStreamError>
+where
+    InternalStreamError: std::error::Error + Send + Sync + 'static,
+{
+    /// Error, that occurs when `recv()` timeout exceeds
+    #[error("Read message timeout")]
+    ReadTimeout,
+
+    /// Error, that occurs when `send()` timeout exceeds
+    #[error("Send message timeout")]
+    SendTimeout,
+
+    /// Error, indicating that empty message was received
+    #[error("No message")]
+    NoMessage,
+
+    /// Error in internal stream representation (typically WebSocket)
+    ///
+    /// Made without `from` macro because it will break `IrohaVersion` variant conversion
+    #[error("Internal stream error: {err}")]
+    InternalStream {
+        /// Internal stream error representation
+        err: InternalStreamError,
```

Comment:
```suggestion
       InternalStreamError,
```

Reply:
Use a tuple-enum here. The structure doesn't do anything, and it's not convention to have both struct and tuple enum variants in rust.